File: 1. Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dict = {} #hash map
        for i,n in enumerate(nums):
            if n in dict:
                return dict[n],i
            else:
                dict[target-n]=i #dictionary of complements
        
"""
    for index in range(0,len(nums)):
            #if nums[index] > target:
                #continue
            #this speedup function doens't work due to negative numbers
            for index2 in range(index+1,len(nums)):
                #if nums[index2] > target:
                    #continue
                if nums[index]+nums[index2]==target:
                    return [index,index2]
                """

File: HashTable/1. Two Sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        //initialize a hash map
        Map<Integer, Integer> hashmap = new HashMap<>();
        int length = nums.length;

        for (int index = 0; index < length; index++) {
            int complement = target - nums[index];
            //look for the complement actively
            if(hashmap.containsKey(complement)){
                return new int[]{hashmap.get(complement), index};
            }
            else{
                hashmap.put(nums[index], index);
            }
        }
        return new int[]{}; //if somehow there is no solution
    }
}